crds/main.go

```diff
@@ -0,0 +1,266 @@
+	required, _, _ := unstruct.NestedStringSlice(schema, "required")
+
+	propertySpecs := make(map[string]pschema.PropertySpec)
+	for propertyName := range properties {
+		propertySchema, _, _ := unstruct.NestedMap(properties, propertyName)
+		propertyDescription, _, _ := unstruct.NestedString(propertySchema, "description")
+		defaultValue, _, _ := unstruct.NestedFieldNoCopy(propertySchema, "default")
+		propertySpecs[propertyName] = pschema.PropertySpec{
+			TypeSpec:    getTypeSpec(propertySchema, name+strings.Title(propertyName), types),
+			Description: propertyDescription,
+			Default:     defaultValue,
+		}
+	}
+
+	types[name+"Args"] = pschema.ObjectTypeSpec{
+		Type:        schemaType,
+		Properties:  propertySpecs,
+		Required:    required,
+		Description: description,
+	}
+}
+
+// getTypeSpec returns the corresponding pschema.TypeSpec for a OpenAPI v3
+// schema. Handles nested pschema.TypeSpecs in case the schema type is of
+// array or object. Also recursively converts and adds all schemas of type
+// object to the types map.
+func getTypeSpec(schema map[string]interface{}, name string, types map[string]pschema.ObjectTypeSpec) pschema.TypeSpec {
+	if schema == nil {
+		return pschema.TypeSpec{}
+	}
+
+	schemaType, foundSchemaType, _ := unstruct.NestedString(schema, "type")
+	if !foundSchemaType {
+		log.Printf("could not find type in %v", schema)
+	}
+
+	if schemaType == "array" {
+		items, _, _ := unstruct.NestedMap(schema, "items")
+		arrayTypeSpec := getTypeSpec(items, name, types)
+		return pschema.TypeSpec{
+			Type:  "array",
+			Items: &arrayTypeSpec,
+		}
+	} else if schemaType == "object" {
+		generateType(schema, name, types)
+		additionalProperties, _, _ := unstruct.NestedMap(schema, "additionalProperties")
+		additionalPropertiesTypeSpec := getTypeSpec(additionalProperties, name, types)
+		return pschema.TypeSpec{
+			Ref:                  "#/types/" + name + "Args",
+			AdditionalProperties: &additionalPropertiesTypeSpec,
+		}
+	} else {
+		return pschema.TypeSpec{
+			Type: schemaType,
+		}
+	}
+}
+
+func generateTypeScriptTypes(types map[string]schema.ObjectTypeSpec) (string, error) {
+	// Create a fake `PackageSpec` that includes the specifed types.
+	// This lets us call the public `nodejs.GeneratePackage` function, which generates code
+	// for an entire package (which isn't really what we want to do, but is existing code
+	// we have with a public entry point, that I'll leverage here for demonstration purposes).
+	// We then return contents of a certain in-memory file that is generated.
+
+	// Note: This is unlikely how we actually want to do the real code gen for the CRD types.
+	// See my comments at the bottom of this function.
+
+	// Create some a property map that references our types, that we can add
+	// as properties to a fake resource, otherwise, our types won't be generated
+	// by the Node.js codegen in `types/input.ts` or `types/output.ts` in-memory files.
```

Comment:
Feel free to pare down the comments that I had written in my example throughout this file.